typescript/src/actions/solana/transfer.ts

```diff
@@ -0,0 +1,228 @@
+ *
+ * @returns The native SOL transfer transaction
+ */
+async function getNativeTransfer({
+  connection,
+  from,
+  to,
+  amount,
+}: GetNativeTransferOptions): Promise<VersionedTransaction> {
+  const { blockhash } = await connection.getLatestBlockhash();
+
+  const instructions = [
+    SystemProgram.transfer({
+      fromPubkey: new PublicKey(from),
+      toPubkey: new PublicKey(to),
+      lamports: amount,
+    }),
+  ];
+
+  const messageV0 = new TransactionMessage({
+    payerKey: new PublicKey(from),
+    recentBlockhash: blockhash,
+    instructions,
+  }).compileToV0Message();
+
+  return new VersionedTransaction(messageV0);
+}
+
+type GetSplTransferOptions = GetNativeTransferOptions & {
+  /**
+   * The mint address of the token
+   */
+  mintAddress: string;
+};
+
+/**
+ * Get the instructions for a SPL token transfer
+ *
+ * @param options - The options for the SPL token transfer
+ *
+ * @param options.connection - The Solana connection
+ * @param options.from - The source address
+ * @param options.to - The destination address
+ * @param options.mintAddress - The mint address of the token
+ * @param options.amount - The amount to transfer
+ *
+ * @returns The SPL token transfer transaction
+ */
+async function getSplTransfer({
+  connection,
+  from,
+  to,
+  mintAddress,
+  amount,
+}: GetSplTransferOptions): Promise<VersionedTransaction> {
+  const fromPubkey = new PublicKey(from);
+  const toPubkey = new PublicKey(to);
+  const mintPubkey = new PublicKey(mintAddress);
+
+  let mintInfo: Awaited<ReturnType<typeof getMint>>;
+  try {
+    mintInfo = await getMint(connection, mintPubkey);
+  } catch (error) {
+    throw new Error(`Failed to fetch mint info for mint address ${mintAddress}. Error: ${error}`);
+  }
+
+  const sourceAta = await getAssociatedTokenAddress(mintPubkey, fromPubkey);
+  const destinationAta = await getAssociatedTokenAddress(mintPubkey, toPubkey);
+
+  const instructions: TransactionInstruction[] = [];
+
+  const sourceAccount = await getAccount(connection, sourceAta);
+  if (sourceAccount.amount < amount) {
+    throw new Error(`Insufficient token balance. Have ${sourceAccount.amount}, need ${amount}`);
+  }
+
+  // Check if destination account exists, if not create it
+  try {
+    await getAccount(connection, destinationAta);
+  } catch {
+    instructions.push(
+      createAssociatedTokenAccountInstruction(fromPubkey, destinationAta, toPubkey, mintPubkey),
+    );
+  }
+
+  instructions.push(
+    createTransferCheckedInstruction(
+      sourceAta,
+      mintPubkey,
+      destinationAta,
+      fromPubkey,
+      amount,
+      mintInfo.decimals,
+    ),
+  );
+
+  return new VersionedTransaction(
+    MessageV0.compile({
+      payerKey: fromPubkey,
+      instructions: instructions,
+      recentBlockhash: (await connection.getLatestBlockhash()).blockhash,
+    }),
+  );
+}
+
+const USDC_MAINNET_MINT = "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v";
```

Comment:
Feels like these should be defined in a constant file, or at least at the top of the file?

Also USDC_MAINNET_MINT_ADDRESS

Last, what happens if the Connection is testnet?

Reply:
> Feels like these should be defined in a constant file, or at least at the top of the file?

Moved to constants.ts

> Also USDC_MAINNET_MINT_ADDRESS

👍 

> Last, what happens if the Connection is testnet?

I updated the fn to throw an error – it seems there is no USDC deployed on testnet. We may want to remove "support" for the Solana testnet, as that's used more for testing the chain itself, whereas devnet is used for application testing

Reply:
+1 on devnet vs testnet